### ISSUE TITLE
Adding an "on" date filter without a specific time to a custom date column returns no results

### ISSUE THREAD

**Issue by daltojohnso** [metabase/metabase#17807]
**Describe the bug**
Adding an "on" date filter without a time to a custom date column shows no results. Filtering on the exact date value works, and filtering using the "between" option works, but you can't filter on a date like `10/07/2019` and see rows with values like `10/07/2019 01:34am`

**Logs**
Here's an example query we send to the `/api/dataset` endpoint. See repro steps below.
```
{"database":1,"query":{"source-table":3,"expressions":{"CC Birth Date":["field",24,null],"CC Created At":["field",18,null]},"fields":[["field",30,null],["field",21,null],["field",23,null],["field",19,null],["field",22,null],["expression","CC Created At"],["field",29,null],["field",20,null],["field",25,null],["field",26,null],["field",24,{"temporal-unit":"default"}],["field",28,null],["field",27,null],["field",18,{"temporal-unit":"default"}],["expression","CC Birth Date"]],"filter":["=",["expression","CC Created At"],"2017-10-07"]},"type":"query","parameters":[]}
```

**To Reproduce**
1. Ask a Question > Custom Question > People dataset
2. Add a custom column that uses the expression `[Created At]`. Name it `CC Created At`
<img width="592" alt="Screen Shot 2021-09-08 at 11 17 58 AM" src="https://user-images.githubusercontent.com/13057258/132564115-1214363f-8dd5-4e5b-9fe5-8fdd76ae281e.png">

3. Run the query. Click on a cell in the "CC Created At" column and select the `=` filter. You should see a filter like "Created At is October 7, 2017 01:34 am" above the table and the table should contain a single result row.

<img width="980" alt="Screen Shot 2021-09-08 at 11 16 12 AM" src="https://user-images.githubusercontent.com/13057258/132564395-af5a230b-eaa2-420b-925d-6390ac6a9cf5.png">

4. Click on the filter to modify it. Add the same date to it but without a time. Run the query and observe that there are now no results.

<img width="989" alt="Screen Shot 2021-09-08 at 11 16 26 AM" src="https://user-images.githubusercontent.com/13057258/132564569-778faede-63a7-4d18-a7dc-2959490316a5.png">

5. You can do this using the original "Created At" field and it'll work correctly.

**Expected behavior**
You'd expect custom column date fields to filter identically to how normal date fields filter.

**Information about your Metabase Installation:**
This is on master, HEAD at `0b60ff9234797cb465dbf4040e2d180120b2fbc1`

**Severity**
P2

**Comment by NickeZ** [comment]
I found out that filtering on a custom field generates the following broken sql:

```
WHERE
 `source`.`Timestamp Stockholm` = DATE(NOW(6))
```

whereas a regular field generates

```
WHERE
      (`table`.`timestamp` >= DATE(NOW(6)))
     
   AND (
        `table`.`timestamp` < DATE(DATE_ADD(NOW(6), INTERVAL 1 day))
      )
```

which is quite different.

**Comment by camsaul** [comment]
This is working as expected on `master`.

<img width="1412" alt="image" src="https://github.com/metabase/metabase/assets/1455846/579cdfb6-9a57-4113-8d85-b0a82164ff9d">

Generated MBQL query:

```clj
{:query
 {:source-query
  {:source-table 50784, :expressions {"CC Created At" [:field 215766 {:base-type :type/DateTimeWithLocalTZ}]}},
  :filter [:= [:field "CC Created At" {:base-type :type/DateTimeWithLocalTZ}] "2017-10-07"]},
 :type :query,
 :database 6792}
```

Compiled SQL:

```sql
Native Query =
SELECT
  "source"."ID" AS "ID",
  "source"."ADDRESS" AS "ADDRESS",
  "source"."EMAIL" AS "EMAIL",
  "source"."PASSWORD" AS "PASSWORD",
  "source"."NAME" AS "NAME",
  "source"."CITY" AS "CITY",
  "source"."LONGITUDE" AS "LONGITUDE",
  "source"."STATE" AS "STATE",
  "source"."SOURCE" AS "SOURCE",
  "source"."BIRTH_DATE" AS "BIRTH_DATE",
  "source"."ZIP" AS "ZIP",
  "source"."LATITUDE" AS "LATITUDE",
  "source"."CREATED_AT" AS "CREATED_AT",
  "source"."CC Created At" AS "CC Created At"
FROM
  (
    SELECT
      "source"."ID" AS "ID",
      "source"."ADDRESS" AS "ADDRESS",
      "source"."EMAIL" AS "EMAIL",
      "source"."PASSWORD" AS "PASSWORD",
      "source"."NAME" AS "NAME",
      "source"."CITY" AS "CITY",
      "source"."LONGITUDE" AS "LONGITUDE",
      "source"."STATE" AS "STATE",
      "source"."SOURCE" AS "SOURCE",
      "source"."BIRTH_DATE" AS "BIRTH_DATE",
      "source"."ZIP" AS "ZIP",
      "source"."LATITUDE" AS "LATITUDE",
      "source"."CREATED_AT" AS "CREATED_AT",
      "source"."CC Created At" AS "CC Created At"
    FROM
      (
        SELECT
          "PUBLIC"."PEOPLE"."ID" AS "ID",
          "PUBLIC"."PEOPLE"."ADDRESS" AS "ADDRESS",
          "PUBLIC"."PEOPLE"."EMAIL" AS "EMAIL",
          "PUBLIC"."PEOPLE"."PASSWORD" AS "PASSWORD",
          "PUBLIC"."PEOPLE"."NAME" AS "NAME",
          "PUBLIC"."PEOPLE"."CITY" AS "CITY",
          "PUBLIC"."PEOPLE"."LONGITUDE" AS "LONGITUDE",
          "PUBLIC"."PEOPLE"."STATE" AS "STATE",
          "PUBLIC"."PEOPLE"."SOURCE" AS "SOURCE",
          "PUBLIC"."PEOPLE"."BIRTH_DATE" AS "BIRTH_DATE",
          "PUBLIC"."PEOPLE"."ZIP" AS "ZIP",
          "PUBLIC"."PEOPLE"."LATITUDE" AS "LATITUDE",
          "PUBLIC"."PEOPLE"."CREATED_AT" AS "CREATED_AT",
          "PUBLIC"."PEOPLE"."CREATED_AT" AS "CC Created At"
        FROM
          "PUBLIC"."PEOPLE"
      ) AS "source"
  ) AS "source"
WHERE
  ("source"."CC Created At" >= ?)
  AND ("source"."CC Created At" < ?)
LIMIT
  1048575
```

**Comment by camsaul** [comment]
So while it appears that the original issue is fixed, this is not working as expected if you add the filter to the same stage as the expression, e.g. 

<img width="323" alt="image" src="https://github.com/metabase/metabase/assets/1455846/29cd01d6-851a-40db-badc-812a76d63f0b">

produces no results.

I will investigate and fix

**Comment by darksciencebase** [comment]
we had to revert https://github.com/metabase/metabase/pull/35523 because of https://github.com/metabase/metabase/issues/35582, reopening

**Comment by bshepherdson** [comment]
Should be straightforward - it was fixed but that fix got reverted and apparently was never reapplied.

Might be related to #33528, and they might both be fixed by reapplying #35523.
